experimental/devices/hx711/hx711.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2018 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// Package hx711 implements an interface to the HX711 analog to digital converter.
+//
+// Datasheet: https://www.mouser.com/ds/2/813/hx711_english-1022875.pdf
+package hx711
+
+import (
+	"errors"
+	"time"
+
+	"periph.io/x/periph/conn/gpio"
+)
+
+var (
+	// TimeoutError is returned from Read and ReadAveraged when the ADC took too
+	// long to indicate data was available.
+	TimeoutError = errors.New("timed out waiting for HX711 to become ready")
+)
+
+// InputMode controls the voltage gain and the channel multiplexer on the HX711.
+// Channel A can be used with a gain of 128 or 64, and Channel B can be used
+// with a gain of 32.
+//
+// Changing the InputMode on an HX711 will only take effect *after* the next
+// read.
+type InputMode int
+
+const (
+	CHANNEL_A_GAIN_128 InputMode = 1
+	CHANNEL_A_GAIN_64  InputMode = 3
+	CHANNEL_B_GAIN_32  InputMode = 2
+
+	dataBits         = 24
+	readPollInterval = 50 * time.Millisecond
+)
+
+type HX711 struct {
+	InputMode InputMode
+
+	clk     gpio.PinOut
+	data    gpio.PinIn
+	useEdge bool
+	done    chan struct{}
+}
+
+// New creates a new HX711 device.
+func New(clk gpio.PinOut, data gpio.PinIn) (*HX711, error) {
+	// Try enabling edge detection on the data pin.
+	var useEdge bool
+	if err := data.In(gpio.PullDown, gpio.FallingEdge); err != nil {
+		if err := data.In(gpio.PullDown, gpio.NoEdge); err != nil {
+			return nil, err
+		}
+		useEdge = false
+	} else {
+		useEdge = true
+	}
+
+	clk.Out(gpio.Low)
+
+	return &HX711{
+		InputMode: CHANNEL_A_GAIN_128,
+		clk:       clk,
+		data:      data,
+		useEdge:   useEdge,
+		done:      nil,
+	}, nil
+}
+
+// IsReady returns true if there is data ready to be read from the ADC.
+func (d *HX711) IsReady() bool {
+	return d.data.Read() == gpio.Low
+}
+
+// Read reads a single value from the ADC.  It blocks until the ADC indicates
+// there is data ready for retrieval.  If the ADC doesn't pull its Data pin low
+// to indicate there is data ready before the timeout is reached, TimeoutError
+// is returned.
+func (d *HX711) Read(timeout time.Duration) (int32, error) {
+	if d.useEdge {
+		// If the clock pin supports edge detection, wait for the falling edge that
+		// indicates the ADC has data.
+		if !d.IsReady() {
+			if !d.data.WaitForEdge(timeout) {
+				return 0, TimeoutError
+			}
+		}
+	} else {
+		// If the clock pin doesn't support edge detection just poll every few
+		// milliseconds.
+		startTime := time.Now()
+		for !d.IsReady() {
+			if time.Now().Sub(startTime) > timeout {
+				return 0, TimeoutError
+			}
+			time.Sleep(readPollInterval)
+		}
+	}
+
+	// Shift the 24-bit 2's compliment value.
+	var value uint32
+	for i := 0; i < dataBits; i++ {
+		d.clk.Out(gpio.High)
```

Comment:
Please check for errors.

Reply:
Is it enough to just check for errors from the initial call to .Out?  I've added a check to New.

Reply:
That's a good question. For device drivers I prefer to check unconditionally. In practice the only case where this can happen in practice is when using a FTDI device over USB, e.g. https://periph.io/x/extra/hostextra/d2xx

Reply:
I understand that there's no room for error here, it can be left as-is for now and we can revisit this later.